app/src/org/commcare/dalvik/activities/LoginActivity.java

```diff
@@ -381,40 +391,43 @@ public void keysDoneOther(LoginActivity r, HttpCalloutOutcomes outcome) {
                         Logger.log(AndroidLogger.TYPE_USER, "unknown");
                         r.raiseLoginMessage(StockMessages.Restore_Unknown, true);
                         break;
+                    case IncorrectPin:
+                        Logger.log(AndroidLogger.TYPE_USER, "incorrect pin");
+                        r.raiseLoginMessage(StockMessages.Auth_InvalidPin, true);
+                        break;
                     default:
                         break;
-                    }
-                }
-            }) {
-                @Override
-                protected void deliverUpdate(LoginActivity receiver, String... update) {
-                    receiver.updateProgress(update[0], TASK_KEY_EXCHANGE);
                 }
-            };
-
-            task.connect(this);
-            task.execute();
+            }
+        };
+    }
 
-            return true;
-        }catch (Exception e) {
-            e.printStackTrace();
-            return false;
+    private int getMatchingUsersCount(String username) {
+        int count = 0;
+        for (UserKeyRecord record : storage()) {

```

Comment:
I would take this as an opportunity to remove the `storage()` and `storage` class field, which only used here. I say this only because I'm on a quest to remove that stupid db close error we get sometimes and have resorted to random tactics like removing as many unneeded references to db objects, to try and prevent ref leaks...